File: agent/src/main/java/fr/cerostudio/remap/ProguardToTiny.java
package fr.cerostudio.remap;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

public class ProguardToTiny {

    private static final Map<String, String> PRIMITIVES = new HashMap<>();
    static {
        PRIMITIVES.put("void", "V");
        PRIMITIVES.put("boolean", "Z");
        PRIMITIVES.put("byte", "B");
        PRIMITIVES.put("char", "C");
        PRIMITIVES.put("short", "S");
        PRIMITIVES.put("int", "I");
        PRIMITIVES.put("long", "J");
        PRIMITIVES.put("float", "F");
        PRIMITIVES.put("double", "D");
    }

    private static final Pattern CLASS_LINE =
        Pattern.compile("^([\\w.$]+)\\s*->\\s*([\\w.$]+):\\s*$");

    private static final Pattern FIELD_LINE =
        Pattern.compile("^\\s+([\\w.$\\[\\]]+)\\s+([\\w$]+)\\s*->\\s*([\\w$]+)\\s*$");

    private static final Pattern METHOD_LINE =
        Pattern.compile("^\\s+(?:\\d+:\\d+:)?([\\w.$\\[\\]]+)\\s+([\\w$<>]+)\\(([^)]*)\\)\\s*->\\s*([\\w$]+)\\s*$");

    public static void main(String[] args) throws IOException {
        if (args.length < 2) {
            System.err.println("Usage: ProguardToTiny <input.proguard.txt> <output.tiny>");
            System.exit(1);
        }

        Path input = Paths.get(args[0]);
        Path output = Paths.get(args[1]);

        List<String> lines = Files.readAllLines(input, java.nio.charset.StandardCharsets.UTF_8);

        Map<String, String> classDeobfToObf = new HashMap<>();
        for (String line : lines) {
            if (line.startsWith("#")) continue;
            Matcher m = CLASS_LINE.matcher(line);
            if (m.matches()) {
                classDeobfToObf.put(m.group(1), m.group(2));
            }
        }

        StringBuilder sb = new StringBuilder();
        sb.append("v1\tofficial\tnamed\n");

        String currentObfClass = null;
        String currentDeobfClass = null;

        for (String line : lines) {
            if (line.startsWith("#") || line.trim().isEmpty()) continue;

            Matcher cm = CLASS_LINE.matcher(line);
            if (cm.matches()) {
                currentDeobfClass = cm.group(1);
                currentObfClass = cm.group(2);
                sb.append("CLASS\t")
                  .append(currentObfClass.replace('.', '/')).append('\t')
                  .append(currentDeobfClass.replace('.', '/')).append('\n');
                continue;
            }

            if (currentObfClass == null) continue;

            Matcher fm = FIELD_LINE.matcher(line);
            if (fm.matches() && !line.contains("(")) {
                String type = fm.group(1);
                String deobfName = fm.group(2);
                String obfName = fm.group(3);
                String desc = toDescriptor(type, classDeobfToObf);
                sb.append("FIELD\t")
                  .append(currentObfClass.replace('.', '/')).append('\t')
                  .append(desc).append('\t')
                  .append(obfName).append('\t')
                  .append(deobfName).append('\n');
                continue;
            }

            Matcher mm = METHOD_LINE.matcher(line);
            if (mm.matches()) {
                String retType = mm.group(1);
                String deobfName = mm.group(2);
                String paramsRaw = mm.group(3);
                String obfName = mm.group(4);

                String desc = buildMethodDescriptor(retType, paramsRaw, classDeobfToObf);
                sb.append("METHOD\t")
                  .append(currentObfClass.replace('.', '/')).append('\t')
                  .append(desc).append('\t')
                  .append(obfName).append('\t')
                  .append(deobfName).append('\n');
            }
        }

        Files.write(output, sb.toString().getBytes(java.nio.charset.StandardCharsets.UTF_8));
        System.out.println("[ProguardToTiny] OK -> " + output);
    }

    private static String buildMethodDescriptor(String retType, String paramsRaw,
                                                  Map<String, String> classMap) {
        StringBuilder params = new StringBuilder("(");
        if (!paramsRaw.trim().isEmpty()) {
            for (String p : paramsRaw.split(",")) {
                params.append(toDescriptor(p.trim(), classMap));
            }
        }
        params.append(")").append(toDescriptor(retType, classMap));
        return params.toString();
    }

    private static String toDescriptor(String type, Map<String, String> classMap) {
        int arrayDepth = 0;
        while (type.endsWith("[]")) {
            arrayDepth++;
            type = type.substring(0, type.length() - 2);
        }

        String desc;
        if (PRIMITIVES.containsKey(type)) {
            desc = PRIMITIVES.get(type);
        } else {
            String obf = classMap.getOrDefault(type, type);
            desc = "L" + obf.replace('.', '/') + ";";
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < arrayDepth; i++) sb.append('[');
        sb.append(desc);
        return sb.toString();
    }
}

File: agent/src/main/scala/fr/cerostudio/remap/ProguardToTiny.scala
package fr.cerostudio.remap

import java.nio.charset.StandardCharsets
import java.nio.file._
import scala.jdk.CollectionConverters._
import scala.collection.mutable
import scala.util.matching.Regex

/**
 * Convertisseur Proguard → Tiny v1.
 *
 * Version Scala utilisant :
 *   - Regex compilées comme vals
 *   - Map immutable pour les primitives
 *   - Pattern matching sur les matchers
 *   - StringBuilder avec interpolation
 */
object ProguardToTiny {

  /** Mapping type Java → descripteur JVM */
  private val Primitives: Map[String, String] = Map(
    "void"    -> "V",
    "boolean" -> "Z",
    "byte"    -> "B",
    "char"    -> "C",
    "short"   -> "S",
    "int"     -> "I",
    "long"    -> "J",
    "float"   -> "F",
    "double"  -> "D"
  )

  // ── Regex de parsing Proguard ───────────────────────────────────────

  private val ClassLineRegex: Regex =
    """^([\w.$]+)\s*->\s*([\w.$]+):\s*$""".r

  private val FieldLineRegex: Regex =
    """^\s+([\w.$\[\]]+)\s+([\w$]+)\s*->\s*([\w$]+)\s*$""".r

  private val MethodLineRegex: Regex =
    """^\s+(?:\d+:\d+:)?([\w.$\[\]]+)\s+([\w$<>]+)\(([^)]*)\)\s*->\s*([\w$]+)\s*$""".r

  def main(args: Array[String]): Unit = {
    if (args.length < 2) {
      System.err.println("Usage: ProguardToTiny <input.proguard.txt> <output.tiny>")
      System.exit(1)
    }

    val input  = Paths.get(args(0))
    val output = Paths.get(args(1))

    try {
      convert(input, output)
      System.out.println(s"[ProguardToTiny] OK -> $output")
    } catch {
      case e: Exception =>
        System.err.println(s"[ProguardToTiny] Erreur: ${e.getMessage}")
        e.printStackTrace()
        System.exit(1)
    }
  }

  def convert(input: Path, output: Path): Unit = {
    val lines = Files.readAllLines(input, StandardCharsets.UTF_8).asScala.toList

    // 1ère passe : construire la table classDeobfToObf
    val classDeobfToObf = mutable.HashMap[String, String]()
    for (line <- lines; if !line.startsWith("#")) {
      line match {
        case ClassLineRegex(deobf, obf) => classDeobfToObf(deobf) = obf
        case _ => // Ignorer
      }
    }

    // 2ème passe : générer le Tiny
    val sb = new StringBuilder()
    sb.append("v1\tofficial\tnamed\n")

    var currentObfClass: String = null
    var currentDeobfClass: String = null

    for (line <- lines; if !line.startsWith("#") && line.trim.nonEmpty) {
      line match {
        case ClassLineRegex(deobf, obf) =>
          currentDeobfClass = deobf
          currentObfClass = obf
          sb.append("CLASS\t")
            .append(obf.replace('.', '/')).append('\t')
            .append(deobf.replace('.', '/')).append('\n')

        case FieldLineRegex(fieldType, deobfName, obfName) if currentObfClass != null && !line.contains("(") =>
          val desc = toDescriptor(fieldType, classDeobfToObf.toMap)
          sb.append("FIELD\t")
            .append(currentObfClass.replace('.', '/')).append('\t')
            .append(desc).append('\t')
            .append(obfName).append('\t')
            .append(deobfName).append('\n')

        case MethodLineRegex(retType, deobfName, paramsRaw, obfName) if currentObfClass != null =>
          val desc = buildMethodDescriptor(retType, paramsRaw, classDeobfToObf.toMap)
          sb.append("METHOD\t")
            .append(currentObfClass.replace('.', '/')).append('\t')
            .append(desc).append('\t')
            .append(obfName).append('\t')
            .append(deobfName).append('\n')

        case _ => // Ignorer les lignes non-matchées
      }
    }

    Files.write(output, sb.toString.getBytes(StandardCharsets.UTF_8))
  }

  // ── Construction des descripteurs ───────────────────────────────────

  private def buildMethodDescriptor(
    retType: String,
    paramsRaw: String,
    classMap: Map[String, String]
  ): String = {
    val params = new StringBuilder("(")
    if (paramsRaw.trim.nonEmpty) {
      for (p <- paramsRaw.split(",")) {
        params.append(toDescriptor(p.trim, classMap))
      }
    }
    params.append(")").append(toDescriptor(retType, classMap))
    params.toString
  }

  private def toDescriptor(typeStr: String, classMap: Map[String, String]): String = {
    // Gérer les tableaux (ex. "int[]", "String[][]")
    var t = typeStr
    var arrayDepth = 0
    while (t.endsWith("[]")) {
      arrayDepth += 1
      t = t.substring(0, t.length - 2)
    }

    val desc = Primitives.getOrElse(
      t,
      "L" + classMap.getOrElse(t, t).replace('.', '/') + ";"
    )

    "[" * arrayDepth + desc
  }
}
